astropy/coordinates/angles/core.py

```diff
@@ -190,6 +190,12 @@ def __new__(cls, angle, unit=None, dtype=np.inexact, copy=True, **kwargs):
 
     @staticmethod
     def _convert_unit_to_angle_unit(unit):
+        for other in (u.rad, u.deg, u.arcmin, u.arcsec, u.hourangle):
+            # hot path: lazily compare object identity to return as quickly as
+            # possible (unit comparison is expensive)
+            if unit is other:
+                return unit
```

Comment:
I'm surprised that this is faster than
```python
if other in (u.rad, u.arcmin, u.arcsec, u.hourangle):
    return other
```

Reply:
I confirm that it is. Here's what I get with your suggestion
```python
In [2]: %timeit Angle(values, u.deg)
5.05 µs ± 31.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
The explanation is that `x in (...)` uses object comparison (`__eq__`) under the hood, so, while it's definitely faster than the current state of the main branch, there's just no beating identity comparison :)

Reply:
Just out of curiosity would:
```python
if other is u.rad or other is u.arcmin or other is u.arcsec or other is u.hourangle:
    return other
``` 
(even though it is horrible to look at) be faster. It might eliminate the cost of setting up the loop.

Reply:
To avoid the comparison loss, one could do use a `set` instead of a `tuple`,
```
unit in {u.rad, u.deg, u.arcmin, u.arcsec, u.hourangle}
```
which is very fast.

Reply:
I tried both your suggestions, and regret to say mine still comes out first 🙈

Reply:
Setting up a set is takes slightly longer than a tuple, so even though inclusion check is fastest, it doesn't seem to compensate for it.

I don't know why @WilliamJamieson's version isn't actually faster than mine !

Reply:
One more thing worth trying would be something like
```
from astropy.units.physical import angle
...
if unit.physical_type is angle:
    return unit
...
```
This might not be the fastest solution, but it's still fast and avoids having to write out the possible angle units.

Reply:
I must say @mhvk 's daring proposal to allow a very small API change with
```patch
diff --git a/astropy/coordinates/angles/core.py b/astropy/coordinates/angles/core.py
index c234100824..489b0fd3c5 100644
--- a/astropy/coordinates/angles/core.py
+++ b/astropy/coordinates/angles/core.py
@@ -190,7 +190,7 @@ class Angle(SpecificTypeQuantity):
 
     @staticmethod
     def _convert_unit_to_angle_unit(unit):
-        return u.hourangle if unit == u.hour else unit
+        return u.hourangle if unit is u.hour else unit
 
     def _set_unit(self, unit):
         super()._set_unit(self._convert_unit_to_angle_unit(unit))
```
takes the cake with
```python
In [1]: %timeit Angle(values, u.deg)
2.82 µs ± 32.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
@MridulS 's benchmark stuff would have been nice here... 👀

Reply:
The `set` one can of course set up outside the function... But I think we're perhaps trying too hard - the rest of the initialization is not that fast that a few 10s of ns are going to matter; the main thing is to get rid of the `==`.

Reply:
Right. So the way I see it, the choice should be between the previous patch I posted in this thread, which is by far the simplest solution *and* the most efficient (but breaks two tests), or something like this, which is about 10% slower (as @mhvk said this should not matter but 100% backward compatible):
```patch
diff --git a/astropy/coordinates/angles/core.py b/astropy/coordinates/angles/core.py
index c234100824..0e20faa3bc 100644
--- a/astropy/coordinates/angles/core.py
+++ b/astropy/coordinates/angles/core.py
@@ -65,6 +65,8 @@ class signed_dms_tuple(NamedTuple):
     """The second value."""
 
 
+_ANGULAR_UNITS = frozenset({u.rad, u.deg, u.arcmin, u.arcsec, u.hourangle})
+
 class Angle(SpecificTypeQuantity):
     """
     One or more angular value(s) with units equivalent to radians or degrees.
@@ -190,6 +192,9 @@ class Angle(SpecificTypeQuantity):
 
     @staticmethod
     def _convert_unit_to_angle_unit(unit):
+        if unit in _ANGULAR_UNITS:
+            return unit
```
```python
In [1]: %timeit Angle(values, u.deg)
3.01 µs ± 56.6 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

Reply:
Since this patch is essentially the same idea but more elegant than what I have at the moment, I'll push it now.

Reply:
What if we keep the function body exactly as it was but just add caching?